Samples/DirectMLNpuInference/main.cpp

```diff
@@ -13,11 +13,47 @@
 
 using Microsoft::WRL::ComPtr;
 
-void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** commandQueueOut, IDMLDevice** dmlDeviceOut) {
-    // Whether to skip adapters which support Graphics in order to target NPU for testing
-    bool forceComputeOnlyDevice = true;
-    bool forceGenericMLDevice = false;
-    
+bool TryGetProperty(IDXCoreAdapter* adapter, DXCoreAdapterProperty prop, std::string& outputValue)
+{
+    if (adapter->IsPropertySupported(prop))
+    {
+        size_t propSize;
+        THROW_IF_FAILED(adapter->GetPropertySize(prop, &propSize));
+
+        outputValue.resize(propSize);
+        THROW_IF_FAILED(adapter->GetProperty(prop, propSize, outputValue.data()));
+
+        // Trim any trailing nul characters. 
+        while (!outputValue.empty() && outputValue.back() == '\0')
+        {
+            outputValue.pop_back();
+        }
+
+        return true;
+    }
+    return false;
+}
+
+// Returns nullptr if not found.
+void GetNonGraphicsAdapter(IDXCoreAdapterList* adapterList, IDXCoreAdapter** outAdapter)
+{
+    for (uint32_t i = 0, adapterCount = adapterList->GetAdapterCount(); i < adapterCount; i++)
+    {
+        ComPtr<IDXCoreAdapter> possibleAdapter;
+        THROW_IF_FAILED(adapterList->GetAdapter(i, IID_PPV_ARGS(&possibleAdapter)));
+
+        if (!possibleAdapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS))
+        {
+            *outAdapter = possibleAdapter.Detach();
+            return;
+        }
+    }
+    *outAdapter = nullptr;
+}
+
+void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** commandQueueOut, IDMLDevice** dmlDeviceOut)
+{
+    // Create Adapter Factory
     ComPtr<IDXCoreAdapterFactory> factory;
     HMODULE dxCoreModule = LoadLibraryW(L"DXCore.dll");
```

Comment:
This is a preexisting issue, but all these HMODULES are leaking. Normally you should explicitly free these or use a smart pointer like wil::unique_hmodule: https://github.com/microsoft/wil/blob/6f60a1b76fb812c6af5db1bc7abdec0001edd43f/include/wil/resource.h#L2687. 

You would need to refactor this sample to not have local variables owning the HMODULEs to really clean this up. Since this is an existing bug, I'm fine with completing this PR and addressing it in a subsequent change. Can you create an internal bug and maybe add a comment next to each LoadLibrary that this is for sample purposes only?

Reply:
Added comment and an internal bug assigned to me tracking this, thanks! :)